doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+
+### Personality to Portlayer Communication
+
+Since all 4 of these must be supported we will require both distinctions for when the container is off and which endpoint to call to distinguish between *to* and *from* operations. 
+
+The proposed solution is to have two endpoints designed for the portlayer api swagger server. Both endpoints will exist along the same request path as two different verbs. The request path should be as such : 
+
+```
+/container/filesystem?ID=
+```
+
+For a copy *to* the filesystem of a container we will do a `PUT` operation
+
+```
+/contrainer/filesystem?ID= PUT
+
+query param:
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - container filesystem target for the copy operation
+
+Response Model :
+
+* stream - tar stream of the data to be moved to the container
+
+* overwriteDir - flag for whether or not we are directed to overwrite existing directories 
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+For a copy *from* the filesystem of a container we will do a `GET` operation. 
+
+```
+/container/filesystem?ID=
+
+Query Param: 
+
+* ID - this is the target container
+
+Request Model : 
+
+* path - target path for the copy operation
+
+* dataStream - tar stream of the data to be copies into the filesystem of the container
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. Each of those endpoints will call their respective portlayer counterparts. There will be several checks during this process. 
+
+Error codes expected by docker : 
+
+```
+200 
+
+400
+
+403 - in the event of a readonly filesystem
+
+404
+
+500
+
+```
+
+Because we need to prevent a poweroff(unless forced) of a container during a copy, we will pass the ID directly to the portlayer rather than fetching a handle. 
+
+Docker does not do much in the way of string checking as far as the path is concerned. This is likely not a big concern for the personality as the actual copy commmand will fail if a bad path string is provided. 
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
+
+## Portlayer Design
+
+ There will be two situations to be concerned with in the portlayer. When the container is *on* and when the container is *off*.
+ 
+ 
+### Off Container Scenario
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/4105)
```

Comment:
#4105 is inherently when we *are* using toolbox. Either this is a typo and a different ticket was intended or clarification is required.